notebook/mortgage_e2e_gquant/mortgage_gquant_plugins.py

```diff
@@ -0,0 +1,1260 @@
+        pass
+
+    logger, console_handler = \
+        hijack_dask_distributed_logger(worker, 'mortgage_workflow_runner')
+
+    worker_name = ''
+    if worker is not None:
+        worker_name = 'WORKER {} '.format(worker.name)
+        logger.info(worker_name + 'RUNNING MORTGAGE gQUANT DataframeFlow')
+        logger.info(worker_name + 'NCCL_P2P_DISABLE: {}'.format(
+            os.environ.get('NCCL_P2P_DISABLE')))
+        logger.info(worker_name + 'CUDA_VISIBLE_DEVICES: {}'.format(
+            os.environ.get('CUDA_VISIBLE_DEVICES')))
+
+    # cpu_df_concat_pandas = None
+    final_perf_acq_arrow_concat = None
+    for ii, run_params_dict in enumerate(mortgage_run_params_dict_list):
+        # performance_path = run_params_dict['csvfile_perfdata']
+        # logger.info(worker_name + 'LOADING: {}'.format(performance_path))
+
+        with warnings.catch_warnings():
+            warnings.simplefilter('ignore', category=UserWarning)
+            final_perf_acq_gdf = mortgage_gquant_run(run_params_dict)
+
+        # CONCATENATE DATAFRAMES AS THEY ARE CALCULATED
+
+        # cpu_df_pandas = gpu_df.to_pandas()
+        # if cpu_df_concat_pandas is None:
+        #     cpu_df_concat_pandas = cpu_df_pandas
+        # else:
+        #     cpu_df_concat_pandas = \
+        #         pd.concat([cpu_df_concat_pandas, cpu_df_pandas])
+        #     del(cpu_df_pandas)
+
+        final_perf_acq_arrow = \
+            final_perf_acq_gdf.to_arrow(preserve_index=False)
+        if final_perf_acq_arrow_concat is None:
+            final_perf_acq_arrow_concat = final_perf_acq_arrow
+        else:
+            final_perf_acq_arrow_concat = pa.concat_tables([
+                final_perf_acq_arrow_concat, final_perf_acq_arrow])
+
+        del(final_perf_acq_gdf)
+        logger.info(worker_name + 'LOADED {} FRAMES'.format(ii+1))
+
+    print_ram_usage(worker_name, logger)
+    logger.info(
+        worker_name +
+        'RUN PYTHON GARBAGE COLLECTION TO MAYBE CLEAR CPU AND GPU MEMORY')
+
+    gc.collect()
+    print_ram_usage(worker_name, logger)
+
+    # df_concat = cpu_df_concat_pandas
+    # delinq_df = df_concat[['delinquency_12']]
+    # indexes_besides_delinq = \
+    #     df_concat.columns.difference(['delinquency_12'])
+    # mortgage_feat_df = df_concat[list(indexes_besides_delinq)]
+    # del(df_concat)
+
+    logger.info(worker_name + 'USING ARROW')
+    # sleep(5)
+
+    cpu_df_concat_arrow = final_perf_acq_arrow_concat
+    delinq_arrow_col = cpu_df_concat_arrow.column('delinquency_12')
+    mortgage_feat_arrow_table = cpu_df_concat_arrow.drop(['delinquency_12'])
+
+    # logger.info(worker_name + 'ARROW TO CUDF')
+    # delinq_arrow_table = pa.Table.from_arrays([delinq_arrow_col])
+    # delinq_df = cudf.DataFrame.from_arrow(delinq_arrow_table)
+    # mortgage_feat_df = cudf.DataFrame.from_arrow(mortgage_feat_arrow_table)
+
+    logger.info(worker_name + 'ARROW TO PANDAS')
+    delinq_df = delinq_arrow_col.to_pandas()
```

Comment:
why need to convert it to pandas ? cudf dataframe cannot be converted to DMatrix?

Reply:
DMatrix can be instantiated from cudf dataframe that's not the problem. The problem is there is not enough memory on GPU to do that. Say if the GPU is 16GB, if you already have a large cudf dataframe 8GB on GPU, instantiating a DMatrix will allocate maybe another 8 or more GB. Very likely you'll get OOM (out of memory) error when instantiating a DMatrix. Therefore to handle large dataframes instantiating a DMatrix from host memory (pandas dataframe) is better. When xgboost training is specified to be on GPU the dmatrix is copied to GPU.

By the way, instantiating a DMatrix directly from dataframes (pandas or cudf dataframes) I believe is currently only supported in rapidai fork of xgboost.

Reply:
this is really an issue that RAPIDS should fix. passing the reference is way more efficient. Maybe file an issue to them.

Reply:
That's not so simple. The DMatrix is core to how XGBoost works. I'm not sure it's even possible to refactor DMatrix so that it would just use pointers directly to a cudf dataframe. I did discuss this on slack with them. I'll add you to my conversation.